lib/integrations/awsoidc/deployservice_update.go

```diff
@@ -0,0 +1,220 @@
+/*
+Copyright 2023 Gravitational, Inc.
+
+Licensed under the Apache License, Version 2.0 (the "License");
+you may not use this file except in compliance with the License.
+You may obtain a copy of the License at
+
+	http://www.apache.org/licenses/LICENSE-2.0
+
+Unless required by applicable law or agreed to in writing, software
+distributed under the License is distributed on an "AS IS" BASIS,
+WITHOUT WARRANTIES OR CONDITIONS OF ANY KIND, either express or implied.
+See the License for the specific language governing permissions and
+limitations under the License.
+*/
+
+package awsoidc
+
+import (
+	"context"
+	"fmt"
+	"time"
+
+	"github.com/aws/aws-sdk-go-v2/aws"
+	"github.com/aws/aws-sdk-go-v2/service/ecs"
+	ecsTypes "github.com/aws/aws-sdk-go-v2/service/ecs/types"
+	"github.com/aws/aws-sdk-go/aws/awsutil"
+	"github.com/gravitational/trace"
+
+	"github.com/gravitational/teleport/lib/modules"
+)
+
+// waitDuration specifies the amount of time to wait for a service to become healthy after an update.
+const waitDuration = time.Minute * 5
+
+// UpdateDeployServiceAgents updates the deploy service agents with the specified teleportVersionTag.
+func UpdateDeployServiceAgents(ctx context.Context, clt DeployServiceClient, teleportClusterName, teleportVersionTag string, ownershipTags AWSTags) error {
+	teleportFlavor := teleportOSS
+	if modules.GetModules().BuildType() == modules.BuildEnterprise {
+		teleportFlavor = teleportEnt
+	}
+	teleportImage := fmt.Sprintf("public.ecr.aws/gravitational/%s-distroless:%s", teleportFlavor, teleportVersionTag)
+
+	if err := updateDeployServiceAgent(ctx, clt, teleportClusterName, teleportImage, ownershipTags); err != nil {
+		return trace.Wrap(err)
+	}
+	return nil
+}
+
+func updateDeployServiceAgent(ctx context.Context, clt DeployServiceClient, teleportClusterName, teleportImage string, ownershipTags AWSTags) error {
+	service, err := getManagedService(ctx, clt, teleportClusterName, ownershipTags)
+	if err != nil {
+		return trace.Wrap(err)
+	}
+
+	taskDefinition, err := getManagedTaskDefinition(ctx, clt, aws.ToString(service.TaskDefinition), ownershipTags)
+	if err != nil {
+		return trace.Wrap(err)
+	}
+
+	currentTeleportImage, err := getTaskDefinitionTeleportImage(taskDefinition)
+	if err != nil {
+		return trace.Wrap(err)
+	}
+
+	if currentTeleportImage == teleportImage {
+		return nil
+	}
+
+	registerTaskDefinitionIn, err := generateTaskDefinitionWithImage(taskDefinition, teleportImage, ownershipTags.ToECSTags())
+	if err != nil {
+		return trace.Wrap(err)
+	}
+
+	registerTaskDefinitionOut, err := clt.RegisterTaskDefinition(ctx, registerTaskDefinitionIn)
+	if err != nil {
+		return trace.Wrap(err)
+	}
+
+	// Update service with new task definition
+	_, err = updateServiceOrRollback(ctx, clt, service, registerTaskDefinitionOut.TaskDefinition)
+	if err != nil {
+		// If update failed, then rollback task definition
+		_, rollbackErr := clt.DeregisterTaskDefinition(ctx, &ecs.DeregisterTaskDefinitionInput{
```

Comment:
What happens after rollback? Will we reattempt the upgrade again?

Reply:
If the update fails, it will be reattempted during the next interval (30mins), but only if it is still within the upgrade window or the critical upgrade flag is still enabled.